archive/tar_stream.ts

```diff
@@ -0,0 +1,353 @@
+              size.toString(8).padStart(size < 8 ** 11 ? 11 : 12, "0") +
+              (size < 8 ** 11 ? " " : "") + // size
+              options.mtime.toString(8).padStart(11, "0") + " " + // mtime
+              " ".repeat(8) + // checksum | To be updated later
+              typeflag + // typeflag
+              "\0".repeat(100) + // linkname
+              "ustar\0" + // magic
+              "00" + // version
+              options.uname.padStart(32, "\0") + // uname
+              options.gname.padStart(32, "\0") + // gname
+              options.devmajor.padStart(8, "\0") + // devmajor
+              options.devminor.padStart(8, "\0"), // devminor
+          ),
+          100,
+        );
+        header.set(prefix, 345); // prefix
+        // Update Checksum
+        header.set(
+          encoder.encode(
+            header.reduce((x, y) => x + y).toString(8).padStart(6, "0") + "\0",
+          ),
+          148,
+        );
+        yield header;
+
+        if ("size" in chunk) {
+          let size = 0;
+          for await (const value of chunk.readable) {
+            size += value.length;
+            yield value;
+          }
+          if (chunk.size !== size) {
+            throw new RangeError(
+              "Provided size did not match bytes read from provided readable",
+            );
+          }
+          if (chunk.size % 512) {
+            yield new Uint8Array(512 - size % 512);
+          }
+        }
+      }
+      yield new Uint8Array(1024);
+    }();
+    this.#readable = new ReadableStream({
+      type: "bytes",
+      async pull(controller) {
+        const { done, value } = await gen.next();
+        if (done) {
+          controller.close();
+          return controller.byobRequest?.respond(0);
+        }
+        if (controller.byobRequest?.view) {
+          const buffer = new Uint8Array(controller.byobRequest.view.buffer);
+
+          const size = buffer.length;
+          if (size < value.length) {
+            buffer.set(value.slice(0, size));
+            controller.byobRequest.respond(size);
+            controller.enqueue(value.slice(size));
+          } else {
+            buffer.set(value);
+            controller.byobRequest.respond(value.length);
+          }
+        } else {
+          controller.enqueue(value);
+        }
+      },
+    });
+  }
+
+  /**
+   * The ReadableStream
+   */
+  get readable(): ReadableStream<Uint8Array> {
+    return this.#readable;
+  }
+
+  /**
+   * The WritableStream
+   */
+  get writable(): WritableStream<TarStreamInput> {
+    return this.#writable;
+  }
+}
+
+/**
+ * parsePathname is a function that validates the correctness of the pathname
+ * being provided.
+ * Function will throw if invalid pathname is provided.
+ * The result can be provided instead of the string version to TarStream,
+ * or can just be used to check in advance of creating the Tar archive.
+ */
+export function parsePathname(
```

Comment:
It's exported so people are able to validate their pathname is valid before putting it in the stream. If people don't have a means to do that then the stream will throw an error part way through

Reply:
I mean, throwing while processing the stream seems like reasonable behavior to me.

Reply:
If you're attempting to tar a lot of files, a simple for loop to validate that all the pathnames are valid can save a lot of time and computational power instead of waiting until the stream gets to the last file only for it to be bad, and have the entire progress made now need to be thrown out as a waste. 

Imagine you were uploading a lot of files as a tarball, where the client was making the tarball in the browser as it streamed the upload, after a lot of progress the stream throws an error because one of the filenames is in an invalid form. Would you not want such a website to be able to check that all the filenames were valid before starting the upload?

Reply:
Why is the filename field the only field that has a validator function?

Reply:
There is also `validTarStreamOptions` function to validate the options provided.

Reply:
Ah, I didn't realise. Ok.